[Add] Muse Glimmer (#3495)
huggingface/blog: muse-glimmer.md

```bash
curl -LsSf https://llama.app/install.sh | sh
```

Then you can start the server as follows.

```bash
llama serve meta-models/Muse-Glimmer-30B-GGUF
```

Once the server has started, you can head to localhost:8080 to chat with the built-in WebUI.

TODO: Insert webui video with this model

You can also query the server as follows.

```bash
curl http://localhost:8080/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
        "messages": [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": "Write a limerick about python exceptions"}
        ]
    }'
```

You can also use llama server with coding agents like Pi.

## Speculative Decoding

DFlash uses a lightweight block-diffusion drafter model to provide same output with extra speed-ups in decoding phase. Transformers and llama.cpp ship support for DFlash drafter of Muse Glimmer day-0.

> 

Reviewer: ```suggestion

Below you can see how speculative decoding can speed-up generation in realistic setups. The video shows llama.cpp webui with DFlash on the left and regular generation on the right. 

<video controls width="100%">
  <source src="https://huggingface.co/merve/smol-vision/resolve/main/llama.cpp-spec.mp4" type="video/mp4">
</video>
```

For this release, we ship support for vLLM with transformers backend.

```bash

# tensor parallel serving across 4 GPUs
vllm serve meta-models/Muse-Glimmer-30B --model-impl transformers --tensor-parallel-size 4

# infer
curl -s http://127.0.0.1:8000/v1/chat/completions \
    -H 'Content-Type: application/json' \
    -d '{
      "model": "username/muse-glimmer-hf-v2",
      "messages": [
        {"role": "user", "content": "Explain tensor parallelism briefly."}
      ],
      "temperature": 0.0,
      "max_tokens": 256
    }'

```

## Fine-tuning with TRL

You can use TRL to fine-tune Muse Glimmer using various methods from SFT to Async GRPO. We have run two experiments on bf16 with Hopper-class GPUs with 80GB VRAM each.

| Workload | Practical minimum |
|---|---:|---:|
| Inference / eval, BF16 | 1×80 GB H100 | 1×H100 |  
| LoRA SFT, BF16 | 1×80 GB H100, microbatch 1 + checkpointing |
| Full SFT, BF16 | 8×80 GB H100 with FSDP/ZeRO-3 |
| LoRA GRPO, Transformers rollouts | 1×80 GB H100, but slow/tight |  
| LoRA GRPO, separate vLLM rollout server | 8×H100: 4 rollout + 4 training |
| Full-finetune GRPO | 8 GPUs is usually insufficient |

> As part of this release, we ship an example to fine-tune Muse Glimmer to fine-tune on small split of MolmoWeb dataset. This shows how to make model generate structured outputs and how to fine-tune on images.

Reviewer: ```suggestion
As part of this release, we ship an example to fine-tune [Muse Glimmer on small split of MolmoWeb dataset](https://huggingface.co/merve/smol-vision/blob/main/qlora_click_grounding.ipynb). This shows how to make model generate structured outputs and how to fine-tune on images.

We also experimented with running the model on [OpenCode with AsyncGRPO example](https://github.com/huggingface/trl/blob/main/examples/scripts/openenv/opencode.py). Model shows strong coding capabilities, so we encourage you to try training with coding environments.
```

---
title: "Meta is back with Muse Glimmer: local, agentic, multimodal, and open source" 
thumbnail: /blog/assets/muse-glimmer/thumbnail.png
authors:
- user: merve
- user: burtenshaw
- user: pcuenq
- user: ariG23498
---

# Meta is back with Muse Glimmer: local, agentic, multimodal, and open source!

Great news from the OGs of open source LLMs! Muse Glimmer, released today, is Meta’s new multimodal model, especially designed for local agentic use cases. Distilled from Muse to **30B** parameters, and released under the **Apache 2.0 license**, it’s ideal deploying locally for privacy, reducing costs, or just hacking around. It’s intended for privacy-aware applications such as coding, document analysis, personal assistants, Claw- or Hermes-like setups.

To celebrate, we are shipping with Meta day-0 support in `transformers`, `llama.cpp`, `vLLM`, Inference Endpoints, and other libraries. We built a few cool things and explain our findings in this blog. **Check out the demo’s below for inspiration.**

> 

Reviewer: You can find all Muse Glimmer models [in this collection](https://huggingface.co/collections/meta-models/muse-glimmer).

Reviewer: omg why did it do like that

\<details\>  
\<summary\>Self-optimization prompt\</summary\>

```bash
You are Muse Glimmer acting as an autonomous inference-optimization engineer for your own serving stack.

Goal: maximize valid single-H100 aggregate completion throughput in tokens/second.

Protocol:
1. Establish a correctness-passing baseline.
2. Test one reversible optimization at a time.
3. Keep the prompt, concurrency, sampling, request count, warm-up, and decode length fixed.
4. Reject results that fail correctness or prefix checks.
5. Record every experiment chronologically with its configuration, raw throughput, correctness, and delta.
6. Keep improvements and revert regressions.
7. Stop after six consecutive regressions or when the experiment budget is exhausted.
8. Report the best valid configuration and exact reproduction command.

Create a minimal scientific animation of the results:
- white background;
- raw tokens/second—never normalize;
- one point revealed per experiment;
- connect every point chronologically;
- begin with the lowest valid result;
- stop at the best result;
- export as a GIF.

Never fabricate, interpolate, or count correctness-failing measurements.
```

\</details\>

> [https://huggingface.co/buckets/huggingface/muse-glimmer-assets/resolve/onyx-optimization-progress.gif?download=true](https://huggingface.co/buckets/huggingface/muse-glimmer-assets/resolve/onyx-optimization-progress.gif?download=true)

Reviewer: ```suggestion
[https://huggingface.co/buckets/huggingface/muse-glimmer-assets/resolve/onyx-optimization-progress.gif?download=true](https://huggingface.co/buckets/huggingface/muse-glimmer-assets/resolve/onyx-optimization-progress.gif?download=true)

## Wrapping Up

We are happy to welcome Muse Glimmer in Hugging Face Hub. Try the models in [this collection](https://huggingface.co/collections/meta-models/muse-glimmer) with your local coding setups today!
```

---
title: "Meta is back with Muse Glimmer: local, agentic, multimodal, and open source" 
thumbnail: /blog/assets/muse-glimmer/thumbnail.png
authors:
- user: merve
- user: burtenshaw
- user: pcuenq
- user: ariG23498
---

# Meta is back with Muse Glimmer: local, agentic, multimodal, and open source!

Great news from the OGs of open source LLMs! Muse Glimmer, released today, is Meta’s new multimodal model, especially designed for local agentic use cases. Distilled from Muse to **30B** parameters, and released under the **Apache 2.0 license**, it’s ideal deploying locally for privacy, reducing costs, or just hacking around. It’s intended for privacy-aware applications such as coding, document analysis, personal assistants, Claw- or Hermes-like setups.

> To celebrate, we are shipping with Meta day-0 support in `transformers`, `llama.cpp`, `vLLM`, Inference Endpoints, and other libraries. We built a few cool things and explain our findings in this blog. **Check out the demo’s below for inspiration.**

Reviewer: ```suggestion
To celebrate, we are shipping with Meta day-0 support in `transformers`, `llama.cpp`, `vLLM`, Inference Endpoints, and other libraries. We built a few cool things and explain our findings in this blog. **Check out the demos below for inspiration.**
```

---
title: "Meta is back with Muse Glimmer: local, agentic, multimodal, and open source" 
thumbnail: /blog/assets/muse-glimmer/thumbnail.png
authors:
- user: merve
- user: burtenshaw
- user: pcuenq
- user: ariG23498
---

# Meta is back with Muse Glimmer: local, agentic, multimodal, and open source!

Great news from the OGs of open source LLMs! Muse Glimmer, released today, is Meta’s new multimodal model, especially designed for local agentic use cases. Distilled from Muse to **30B** parameters, and released under the **Apache 2.0 license**, it’s ideal deploying locally for privacy, reducing costs, or just hacking around. It’s intended for privacy-aware applications such as coding, document analysis, personal assistants, Claw- or Hermes-like setups.

To celebrate, we are shipping with Meta day-0 support in `transformers`, `llama.cpp`, `vLLM`, Inference Endpoints, and other libraries. We built a few cool things and explain our findings in this blog. **Check out the demos below for inspiration.**

> 

Reviewer: ```suggestion
You can find all Muse Glimmer models [in this collection](https://huggingface.co/collections/meta-models/muse-glimmer).
```
